Modules/Microsoft365DSC/Modules/M365DSCUtil.psm1

```diff
@@ -3326,6 +3326,10 @@ function Get-M365DSCExportContentForResource
     {
         $primaryKey = $Results.DisplayName
     }
+    elseif ($Results.ContainsKey('Name'))
```

Comment:
Can we make sure, that the Name is a mandatory parameter? I'm wondering if there are any other resource that might use Name a a non-key-parameter.

Reply:
Yes, there's a few resources that use property *Name* and it's not mandatory, but then again the same already happens with *DisplayName* and could happen with any of the other properties.

The only way I see it working properly would be to add an extra parameter to the function where you could define the key but would involve changing all resources.

Reply:
If this is not accepted I can live with that, it's not big of a problem for me but like I said the way the code is today it can cause problems where it's not supposed to happen since it relies on the ordering of the conditions instead of the actual knowledge of what's a key parameter, or not, of each resource.

Reply:
Would it be possible to import the resource in this step and get the keys on the fly?

Reply:
If I’m not mistaken, we do have this kind of functionality within the code. 🔎

Reply:
We could always do something like this, with error handling of course but it works.

```powershell
# $ResourceName is passed as a parameter of Get-M365DSCExportContentForResource
$ResourceName = "IntuneSettingCatalogCustomPolicyWindows10"
(Get-DscResource -Name $ResourceName).Properties.Where({$_.IsMandatory}) | Select-Object -Property Name
```

```
Name
----
Name
```

Reply:
Great find. This would be more generic also for future use.

Reply:
Taking out the usual suspects for keys I still have these resources below that have more than one key and are not even covered by what's in the code today. From these with multiple keys we need to select a single one of them as part of the heuristics, which will be almost the same as what's there already but with *Get-DscResource* helping out.

```code
Name                           Value                                                                                                                                                                               ----                           -----                                                                                                                                                                               Name                           AADRoleEligibilityScheduleRequest                                                                                                                                                   Keys                           {Principal, RoleDefinition}                                                                                                                                                         Name                           M365DSCRuleEvaluation                                                                                                                                                               Keys                           {ResourceName, RuleDefinition}                                                                                                                                                      Name                           SCComplianceSearchAction                                                                                                                                                            Keys                           {Action, SearchName}                                                                                                                                                                Name                           SPOPropertyBag                                                                                                                                                                      Keys                           {Key, Url, Value}                                                                                                                                                                   Name                           SPOSiteAuditSettings                                                                                                                                                                Keys                           {AuditFlags, Url}                                                                                                                                                                   Name                           SPOSiteDesignRights                                                                                                                                                                 Keys                           {Rights, SiteDesignTitle}                                                                                                                                                           Name                           SPOStorageEntity                                                                                                                                                                    Keys                           {Key, SiteUrl}                                                                                                                                                                      Name                           TeamsGroupPolicyAssignment                                                                                                                                                          Keys                           {GroupDisplayName, PolicyType}                                                                                                                                                      Name                           TeamsUser                                                                                                                                                                           Keys                           {TeamName, User}
```

Reply:
I would join them together to a longer string to stay unique.

Reply:
Cool that is better for me than joining with dashes since I also rely on splitting this ResourceInstanceName on blueprints in order to do some stuff.

Will have a look into doing this, I'll update the PR once I have it tested.

Reply:
@andikrueger I'm making a change here for the keys of the resources and now during a full Intune export I have 2 resources that complain about duplicate resources with the same key even when I go back using the original code!!!

I thought I was going crazy but I found out why, this is because of Graph 2.13.1 since it adds the *ResponseHeaders* as an additional object now on some queries and these 2 resources have a condition looking if the $getValue is an array which then throws an exception.

MSFT_IntuneDeviceEnrollmentStatusPageWindows10/MSFT_IntuneDeviceEnrollmentStatusPageWindows10.psm1:152:        if($getValue -is [Array])
MSFT_IntuneWindowsAutopilotDeploymentProfileAzureADJoined/MSFT_IntuneWindowsAutopilotDeploymentProfileAzureADJoined.psm1:130:        if($getValue -is [Array])

As I explained on https://github.com/microsoft/Microsoft365DSC/issues/4341 this is a problem with Graph quality on these past few releases, I really can't use these 2 policies until the issue with Graph is resolved so how can I proceed in not only solving #4333 but these additional issues?

Reply:
Hopefully this week's release includes Graph 2.14.1 which solves the ReturnHeaderValues issue, until then I'll keep this PR as draft.

Reply:
Please have a look at this PR #4348

Reply:
Thank you I'm already aware but unfortunately I don't have time to look at this right at this moment.